go test failure on master (segmentation violation)

<!--
  Come join our community on Slack https://pion.ly/slack for quicker support and more resources

  We also provide a Knowledge Base https://pion.ly/knowledge-base that covers Pion and general WebRTC
  
  We maintain a FAQ of common issues and questions https://pion.ly/knowledge-base/pion-basics/faq/

 Please answer these questions before submitting your issue. Thanks!
-->

### Your environment.
  - Version: 3b470331467299f1cb7d3b0b9a3955e396231521

### What did you do?

Ran `go test`

### What did you expect?

Test to pass

### What happened?


sam@samubuntu:~/Dev/pion/ice$ go test
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x752230]

goroutine 91 [running]:
github.com/pion/ice/v2.(*Agent).handleInbound(0xc0000aea00, 0xc000088460, 0x904580, 0xc0000b0500, 0x8fa660, 0xc00028a030)
	/home/sam/Dev/pion/ice/agent.go:922 +0x800
github.com/pion/ice/v2.handleInboundCandidateMsg.func1(0xc0000aea00)
	/home/sam/Dev/pion/ice/candidate_base.go:113 +0x58
github.com/pion/ice/v2.(*Agent).run(0xc0000aea00, 0xc00044ef10, 0xc000112000, 0xc00028c000, 0x0)
	/home/sam/Dev/pion/ice/agent.go:168 +0x262
github.com/pion/ice/v2.handleInboundCandidateMsg(0x904580, 0xc0000b0500, 0xc00017c000, 0x70, 0x2000, 0x8fa660, 0xc00028a030, 0x902a40, 0xc00007c000)
	/home/sam/Dev/pion/ice/candidate_base.go:112 +0x538
github.com/pion/ice/v2.(*candidateBase).recvLoop(0xc0000b0500)
	/home/sam/Dev/pion/ice/candidate_base.go:97 +0xe1
created by github.com/pion/ice/v2.(*candidateBase).start
	/home/sam/Dev/pion/ice/candidate_base.go:81 +0xe0
exit status 2
FAIL	github.com/pion/ice/v2	4.349s

It occurs with different tests, racily randomly

```
sam@samubuntu:~/Dev/pion/ice$ go version
go version go1.14.2 linux/amd64
sam@samubuntu:~/Dev/pion/ice$ go test -v
=== RUN   TestPairSearch
--- PASS: TestPairSearch (0.00s)
=== RUN   TestPairPriority
--- PASS: TestPairPriority (0.00s)
=== RUN   TestOnSelectedCandidatePairChange
--- PASS: TestOnSelectedCandidatePairChange (0.00s)
=== RUN   TestHandlePeerReflexive
=== RUN   TestHandlePeerReflexive/UDP_pflx_candidate_from_handleInbound()
=== RUN   TestHandlePeerReflexive/Bad_network_type_with_handleInbound()
=== RUN   TestHandlePeerReflexive/Success_from_unknown_remote,_prflx_candidate_MUST_only_be_created_via_Binding_Request
--- PASS: TestHandlePeerReflexive (0.42s)
    --- PASS: TestHandlePeerReflexive/UDP_pflx_candidate_from_handleInbound() (0.00s)
    --- PASS: TestHandlePeerReflexive/Bad_network_type_with_handleInbound() (0.01s)
    --- PASS: TestHandlePeerReflexive/Success_from_unknown_remote,_prflx_candidate_MUST_only_be_created_via_Binding_Request (0.00s)
=== RUN   TestConnectivityOnStartup
--- PASS: TestConnectivityOnStartup (0.40s)
=== RUN   TestConnectivityLite
--- PASS: TestConnectivityLite (2.40s)
=== RUN   TestInboundValidity
=== RUN   TestInboundValidity/Invalid_Binding_requests_should_be_discarded
=== RUN   TestInboundValidity/Invalid_Binding_success_responses_should_be_discarded
=== RUN   TestInboundValidity/Discard_non-binding_messages
=== RUN   TestInboundValidity/Valid_bind_request
=== RUN   TestInboundValidity/Valid_bind_without_fingerprint
=== RUN   TestInboundValidity/Success_with_invalid_TransactionID
--- PASS: TestInboundValidity (0.41s)
    --- PASS: TestInboundValidity/Invalid_Binding_requests_should_be_discarded (0.00s)
    --- PASS: TestInboundValidity/Invalid_Binding_success_responses_should_be_discarded (0.00s)
    --- PASS: TestInboundValidity/Discard_non-binding_messages (0.00s)
    --- PASS: TestInboundValidity/Valid_bind_request (0.00s)
    --- PASS: TestInboundValidity/Valid_bind_without_fingerprint (0.00s)
    --- PASS: TestInboundValidity/Success_with_invalid_TransactionID (0.00s)
=== RUN   TestInvalidAgentStarts
--- PASS: TestInvalidAgentStarts (0.50s)
=== RUN   TestConnectionStateCallback
--- PASS: TestConnectionStateCallback (2.90s)
=== RUN   TestInvalidGather
=== RUN   TestInvalidGather/Gather_with_no_OnCandidate_should_error
--- PASS: TestInvalidGather (0.00s)
    --- PASS: TestInvalidGather/Gather_with_no_OnCandidate_should_error (0.00s)
=== RUN   TestCandidatePairStats
--- PASS: TestCandidatePairStats (0.40s)
=== RUN   TestLocalCandidateStats
--- PASS: TestLocalCandidateStats (0.40s)
=== RUN   TestRemoteCandidateStats
--- PASS: TestRemoteCandidateStats (0.40s)
=== RUN   TestInitExtIPMapping
--- PASS: TestInitExtIPMapping (0.40s)
=== RUN   TestBindingRequestTimeout
--- PASS: TestBindingRequestTimeout (0.40s)
=== RUN   TestAgentCredentials
--- PASS: TestAgentCredentials (0.00s)
=== RUN   TestConnectionStateFailedDeleteAllCandidates
--- PASS: TestConnectionStateFailedDeleteAllCandidates (2.65s)
=== RUN   TestConnectionStateConnectingToFailed
--- PASS: TestConnectionStateConnectingToFailed (2.40s)
=== RUN   TestAgentRestart
=== RUN   TestAgentRestart/Restart_During_Gather
=== RUN   TestAgentRestart/Restart_When_Closed
=== RUN   TestAgentRestart/Restart_One_Side
=== RUN   TestAgentRestart/Restart_Both_Sides
--- PASS: TestAgentRestart (2.45s)
    --- PASS: TestAgentRestart/Restart_During_Gather (0.00s)
    --- PASS: TestAgentRestart/Restart_When_Closed (0.00s)
    --- PASS: TestAgentRestart/Restart_One_Side (2.00s)
    --- PASS: TestAgentRestart/Restart_Both_Sides (0.05s)
=== RUN   TestRelayOnlyConnection
--- PASS: TestRelayOnlyConnection (2.40s)
=== RUN   TestServerReflexiveOnlyConnection
--- PASS: TestServerReflexiveOnlyConnection (2.40s)
=== RUN   TestCandidatePriority
--- PASS: TestCandidatePriority (0.00s)
=== RUN   TestCandidateLastSent
--- PASS: TestCandidateLastSent (0.00s)
=== RUN   TestCandidateLastReceived
--- PASS: TestCandidateLastReceived (0.00s)
=== RUN   TestCandidatePairPriority
--- PASS: TestCandidatePairPriority (0.00s)
=== RUN   TestCandidatePairEquality
--- PASS: TestCandidatePairEquality (0.00s)
=== RUN   TestConnectivityVNet
=== RUN   TestConnectivityVNet/Full-cone_NATs_on_both_ends
=== RUN   TestConnectivityVNet/Symmetric_NATs_on_both_ends
=== RUN   TestConnectivityVNet/1:1_NAT_with_host_candidate_vs_Symmetric_NATs
=== RUN   TestConnectivityVNet/1:1_NAT_with_srflx_candidate_vs_Symmetric_NATs
--- PASS: TestConnectivityVNet (9.40s)
    --- PASS: TestConnectivityVNet/Full-cone_NATs_on_both_ends (3.00s)
    --- PASS: TestConnectivityVNet/Symmetric_NATs_on_both_ends (4.00s)
    --- PASS: TestConnectivityVNet/1:1_NAT_with_host_candidate_vs_Symmetric_NATs (0.00s)
    --- PASS: TestConnectivityVNet/1:1_NAT_with_srflx_candidate_vs_Symmetric_NATs (2.00s)
=== RUN   TestDisconnectedToConnected
--- PASS: TestDisconnectedToConnected (1.44s)
=== RUN   TestWriteUseValidPair
--- PASS: TestWriteUseValidPair (0.42s)
=== RUN   TestExternalIPMapper
=== RUN   TestExternalIPMapper/validateIPString
=== RUN   TestExternalIPMapper/newExternalIPMapper
=== RUN   TestExternalIPMapper/newExternalIPMapper_with_explicit_local_IP
=== RUN   TestExternalIPMapper/newExternalIPMapper_with_inplicit_and_explicit_local_IP
=== RUN   TestExternalIPMapper/findExternalIP_without_explicit_local_IP
=== RUN   TestExternalIPMapper/findExternalIP_with_explicit_local_IP
=== RUN   TestExternalIPMapper/findExternalIP_with_empty_map
--- PASS: TestExternalIPMapper (0.00s)
    --- PASS: TestExternalIPMapper/validateIPString (0.00s)
    --- PASS: TestExternalIPMapper/newExternalIPMapper (0.00s)
    --- PASS: TestExternalIPMapper/newExternalIPMapper_with_explicit_local_IP (0.00s)
    --- PASS: TestExternalIPMapper/newExternalIPMapper_with_inplicit_and_explicit_local_IP (0.00s)
    --- PASS: TestExternalIPMapper/findExternalIP_without_explicit_local_IP (0.00s)
    --- PASS: TestExternalIPMapper/findExternalIP_with_explicit_local_IP (0.00s)
    --- PASS: TestExternalIPMapper/findExternalIP_with_empty_map (0.00s)
=== RUN   TestListenUDP
--- PASS: TestListenUDP (0.00s)
=== RUN   TestSTUNConcurrency
--- PASS: TestSTUNConcurrency (5.20s)
=== RUN   TestTURNConcurrency
=== RUN   TestTURNConcurrency/UDP_Relay
=== RUN   TestTURNConcurrency/TCP_Relay
=== RUN   TestTURNConcurrency/TLS_Relay
=== RUN   TestTURNConcurrency/DTLS_Relay
--- PASS: TestTURNConcurrency (8.02s)
    --- PASS: TestTURNConcurrency/UDP_Relay (0.00s)
    --- PASS: TestTURNConcurrency/TCP_Relay (0.00s)
    --- PASS: TestTURNConcurrency/TLS_Relay (0.01s)
    --- PASS: TestTURNConcurrency/DTLS_Relay (0.00s)
=== RUN   TestCloseConnLog
--- PASS: TestCloseConnLog (0.00s)
=== RUN   TestVNetGather
=== RUN   TestVNetGather/No_local_IP_address
=== RUN   TestVNetGather/Gather_a_dynamic_IP_address
=== RUN   TestVNetGather/listenUDP
--- PASS: TestVNetGather (0.40s)
    --- PASS: TestVNetGather/No_local_IP_address (0.00s)
    --- PASS: TestVNetGather/Gather_a_dynamic_IP_address (0.00s)
    --- PASS: TestVNetGather/listenUDP (0.00s)
=== RUN   TestVNetGatherWithNAT1To1
=== RUN   TestVNetGatherWithNAT1To1/gather_1:1_NAT_external_IPs_as_host_candidates
=== RUN   TestVNetGatherWithNAT1To1/gather_1:1_NAT_external_IPs_as_srflx_candidates
--- PASS: TestVNetGatherWithNAT1To1 (0.40s)
    --- PASS: TestVNetGatherWithNAT1To1/gather_1:1_NAT_external_IPs_as_host_candidates (0.00s)
    --- PASS: TestVNetGatherWithNAT1To1/gather_1:1_NAT_external_IPs_as_srflx_candidates (0.00s)
=== RUN   TestVNetGatherWithInterfaceFilter
=== RUN   TestVNetGatherWithInterfaceFilter/InterfaceFilter_should_exclude_the_interface
=== RUN   TestVNetGatherWithInterfaceFilter/InterfaceFilter_should_not_exclude_the_interface
--- PASS: TestVNetGatherWithInterfaceFilter (0.40s)
    --- PASS: TestVNetGatherWithInterfaceFilter/InterfaceFilter_should_exclude_the_interface (0.00s)
    --- PASS: TestVNetGatherWithInterfaceFilter/InterfaceFilter_should_not_exclude_the_interface (0.00s)
=== RUN   TestConnectedState_String
--- PASS: TestConnectedState_String (0.00s)
=== RUN   TestGatheringState_String
--- PASS: TestGatheringState_String (0.00s)
=== RUN   TestControlled_GetFrom
=== RUN   TestControlled_GetFrom/IncorrectSize
--- PASS: TestControlled_GetFrom (0.00s)
    --- PASS: TestControlled_GetFrom/IncorrectSize (0.00s)
=== RUN   TestControlling_GetFrom
=== RUN   TestControlling_GetFrom/IncorrectSize
--- PASS: TestControlling_GetFrom (0.00s)
    --- PASS: TestControlling_GetFrom/IncorrectSize (0.00s)
=== RUN   TestControl_GetFrom
=== RUN   TestControl_GetFrom/Blank
=== RUN   TestControl_GetFrom/Controlling
=== RUN   TestControl_GetFrom/Controlling/IncorrectSize
=== RUN   TestControl_GetFrom/Controlled
=== RUN   TestControl_GetFrom/Controlled/IncorrectSize
--- PASS: TestControl_GetFrom (0.00s)
    --- PASS: TestControl_GetFrom/Blank (0.00s)
    --- PASS: TestControl_GetFrom/Controlling (0.00s)
        --- PASS: TestControl_GetFrom/Controlling/IncorrectSize (0.00s)
    --- PASS: TestControl_GetFrom/Controlled (0.00s)
        --- PASS: TestControl_GetFrom/Controlled/IncorrectSize (0.00s)
=== RUN   TestMulticastDNSOnlyConnection
--- PASS: TestMulticastDNSOnlyConnection (2.40s)
=== RUN   TestMulticastDNSMixedConnection
--- PASS: TestMulticastDNSMixedConnection (2.40s)
=== RUN   TestMulticastDNSStaticHostName
--- PASS: TestMulticastDNSStaticHostName (0.40s)
=== RUN   TestNetworkTypeParsing_Success
--- PASS: TestNetworkTypeParsing_Success (0.00s)
=== RUN   TestNetworkTypeParsing_Failure
--- PASS: TestNetworkTypeParsing_Failure (0.00s)
=== RUN   TestPriority_GetFrom
=== RUN   TestPriority_GetFrom/IncorrectSize
--- PASS: TestPriority_GetFrom (0.00s)
    --- PASS: TestPriority_GetFrom/IncorrectSize (0.00s)
=== RUN   TestStressDuplex
--- PASS: TestStressDuplex (0.45s)
=== RUN   TestTimeout
    TestTimeout: transport_test.go:54: Connection timed out in 6000.138167 msec
    TestTimeout: transport_test.go:54: Connection timed out in 6000.124216 msec
--- PASS: TestTimeout (12.06s)
=== RUN   TestReadClosed
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x752230]

goroutine 1087 [running]:
github.com/pion/ice/v2.(*Agent).handleInbound(0xc0000aef00, 0xc000453810, 0x904580, 0xc0003a80a0, 0x8fa660, 0xc00016b5c0)
	/home/sam/Dev/pion/ice/agent.go:922 +0x800
github.com/pion/ice/v2.handleInboundCandidateMsg.func1(0xc0000aef00)
	/home/sam/Dev/pion/ice/candidate_base.go:113 +0x58
github.com/pion/ice/v2.(*Agent).run(0xc0000aef00, 0xc000687f10, 0xc000038de0, 0xc00056a150, 0x0)
	/home/sam/Dev/pion/ice/agent.go:168 +0x262
github.com/pion/ice/v2.handleInboundCandidateMsg(0x904580, 0xc0003a80a0, 0xc000412000, 0x70, 0x2000, 0x8fa660, 0xc00016b5c0, 0x902a40, 0xc00033ed00)
	/home/sam/Dev/pion/ice/candidate_base.go:112 +0x538
github.com/pion/ice/v2.(*candidateBase).recvLoop(0xc0003a80a0)
	/home/sam/Dev/pion/ice/candidate_base.go:97 +0xe1
created by github.com/pion/ice/v2.(*candidateBase).start
	/home/sam/Dev/pion/ice/candidate_base.go:81 +0xe0
exit status 2
FAIL	github.com/pion/ice/v2	64.809s
```

Oof maybe I introduced this. Are you able to pin this down to one commit @nerd2 ? I will work on reproducing.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x752230]

goroutine 1087 [running]:
github.com/pion/ice/v2.(*Agent).handleInbound(0xc0000aef00, 0xc000453810, 0x904580, 0xc0003a80a0, 0x8fa660, 0xc00016b5c0)
```

I don't know any easy way to debug this from the log. Not a nil pointer, so I guess `a.selector` is invalid? But we set `a.selector` in a `a.run` so doesn't seem possible to have a racey read/write. @at-wat you are a better debugger then me :)

With trace
```
=== RUN   TestConnectivityOnStartup
ice WARNING: 2020/06/26 10:21:42 vnet is enabled
ice TRACE: 10:21:42.584755 agent.go:433: Set selected candidate pair: <nil>
ice WARNING: 2020/06/26 10:21:42 vnet is enabled
ice TRACE: 10:21:42.585149 agent.go:433: Set selected candidate pair: <nil>
ice DEBUG: 10:21:42.585494 agent.go:339: Started agent: isControlling? true, remoteUfrag: "yzbIoSKFYlBFgfpl", remotePwd: "PlfMeALFcUIrKhjjLfRMvTqoljVsRrbT"
ice INFO: 2020/06/26 10:21:42 Setting new connection state: Checking
ice TRACE: 10:21:42.585669 agent.go:453: pinging all candidates
ice TRACE: 10:21:42.585805 agent.go:779: ping STUN from host 192.168.0.2:5425 to host 192.168.0.1:5540
ice DEBUG: 10:21:42.585994 agent.go:339: Started agent: isControlling? false, remoteUfrag: "dHlcnbHVTULkkjEK", remotePwd: "meZBdkdJLcwNeBImwltSFUGUBhIHZNHN"
ice TRACE: 10:21:42.586140 agent.go:923: inbound STUN (Request) from 192.168.0.2:5425 to host 192.168.0.1:5540
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x751d50]

goroutine 83 [running]:
github.com/pion/ice/v2.(*Agent).handleInbound(0xc0000c8780, 0xc0000cc1e0, 0x9048a0, 0xc0000ea000, 0x8fa980, 0xc00009cf30)
	/home/at-wat/go/src/github.com/pion/ice/agent.go:925 +0x800
github.com/pion/ice/v2.handleInboundCandidateMsg.func1(0xc0000c8780)
	/home/at-wat/go/src/github.com/pion/ice/candidate_base.go:113 +0x58
github.com/pion/ice/v2.(*Agent).run(0xc0000c8780, 0xc00005df10, 0xc00008e0c0, 0xc000314150, 0x1000000000001)
	/home/at-wat/go/src/github.com/pion/ice/agent.go:168 +0x262
github.com/pion/ice/v2.handleInboundCandidateMsg(0x9048a0, 0xc0000ea000, 0xc000214000, 0x70, 0x2000, 0x8fa980, 0xc00009cf30, 0x902d60, 0xc0000df680)
	/home/at-wat/go/src/github.com/pion/ice/candidate_base.go:112 +0x538
github.com/pion/ice/v2.(*candidateBase).recvLoop(0xc0000ea000)
	/home/at-wat/go/src/github.com/pion/ice/candidate_base.go:97 +0xe1
created by github.com/pion/ice/v2.(*candidateBase).start
	/home/at-wat/go/src/github.com/pion/ice/candidate_base.go:81 +0xe0
FAIL	github.com/pion/ice/v2	1.453s
FAIL
```

--- a/agent.go
+++ b/agent.go
@@ -35,9 +35,6 @@ type Agent struct {
 	onSelectedCandidatePairChangeHdlr atomic.Value // func(Candidate, Candidate)
 	onCandidateHdlr                   atomic.Value // func(Candidate)
 
-	// Used to block double Dial/Accept
-	opened bool
-
 	// State owned by the taskLoop
 	onConnected     chan struct{}
 	onConnectedOnce sync.Once
@@ -56,8 +53,9 @@ type Agent struct {
 	mDNSName string
 	mDNSConn *mdns.Conn
 
-	haveStarted   bool
 	muHaveStarted sync.Mutex
+	startedCh     <-chan struct{}
+	startedFn     func()
 	isControlling bool
 
 	maxBindingRequests uint16
@@ -217,6 +215,8 @@ func NewAgent(config *AgentConfig) (*Agent, error) {
 		}
 	}
 
+	startedCtx, startedFn := context.WithCancel(context.Background())
+
 	a := &Agent{
 		tieBreaker:       rand.New(rand.NewSource(time.Now().UnixNano())).Uint64(),
 		lite:             config.Lite,
@@ -229,6 +229,8 @@ func NewAgent(config *AgentConfig) (*Agent, error) {
 		onConnected:      make(chan struct{}),
 		buffer:           packetio.NewBuffer(),
 		done:             make(chan struct{}),
+		startedCh:        startedCtx.Done(),
+		startedFn:        startedFn,
 		chanState:        make(chan ConnectionState, 1),
 		portmin:          config.PortMin,
 		portmax:          config.PortMax,
@@ -329,13 +331,14 @@ func (a *Agent) startOnConnectionStateChangeRoutine() {
 func (a *Agent) startConnectivityChecks(isControlling bool, remoteUfrag, remotePwd string) error {
 	a.muHaveStarted.Lock()
 	defer a.muHaveStarted.Unlock()
-	if a.haveStarted {
+	select {
+	case <-a.startedCh:
 		return ErrMultipleStart
+	default:
 	}
 	if err := a.SetRemoteCredentials(remoteUfrag, remotePwd); err != nil {
 		return err
 	}
-	a.haveStarted = true
 
 	a.startOnConnectionStateChangeRoutine()
 	a.log.Debugf("Started agent: isControlling? %t, remoteUfrag: %q, remotePwd: %q", isControlling, remoteUfrag, remotePwd)
@@ -356,6 +359,7 @@ func (a *Agent) startConnectivityChecks(isControlling bool, remoteUfrag, remoteP
 		}
 
 		a.selector.Start()
+		a.startedFn()
 
 		agent.updateConnectionState(ConnectionStateChecking)
 
@@ -653,7 +657,7 @@ func (a *Agent) addRemoteCandidate(c Candidate) {
 
 func (a *Agent) addCandidate(c Candidate, candidateConn net.PacketConn) error {
 	return a.run(func(agent *Agent) {
-		c.start(a, candidateConn)
+		c.start(a, candidateConn, a.startedCh)
 
 		set := a.localCandidates[c.NetworkType()]
 		for _, candidate := range set {
@@ -715,6 +719,7 @@ func (a *Agent) Close() error {
 		close(agent.done)
 
 		a.deleteAllCandidates()
+		a.startedFn()
 
 		if err := a.buffer.Close(); err != nil {
 			a.log.Warnf("failed to close buffer: %v", err)

--- a/agent_test.go
+++ b/agent_test.go
@@ -228,12 +228,12 @@ func runAgentTest(t *testing.T, config *AgentConfig, task func(a *Agent)) {
 }
 
 func TestHandlePeerReflexive(t *testing.T) {
-	// Limit runtime in case of deadlocks
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 2)
+	defer lim.Stop()
 
 	t.Run("UDP pflx candidate from handleInbound()", func(t *testing.T) {
 		var config AgentConfig
@@ -466,11 +466,11 @@ func TestConnectivityOnStartup(t *testing.T) {
 }
 
 func TestConnectivityLite(t *testing.T) {
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
 
 	stunServerURL := &URL{
 		Scheme: SchemeTypeSTUN,
@@ -710,11 +710,11 @@ func TestInvalidAgentStarts(t *testing.T) {
 
 // Assert that Agent emits Connecting/Connected/Disconnected/Failed/Closed messages
 func TestConnectionStateCallback(t *testing.T) {
-	lim := test.TimeOut(time.Second * 5)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 5)
+	defer lim.Stop()
 
 	disconnectedDuration := time.Second
 	failedDuration := time.Second
@@ -1253,11 +1253,11 @@ func TestAgentCredentials(t *testing.T) {
 // Assert that Agent on Failure deletes all existing candidates
 // User can then do an ICE Restart to bring agent back
 func TestConnectionStateFailedDeleteAllCandidates(t *testing.T) {
-	lim := test.TimeOut(time.Second * 5)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 5)
+	defer lim.Stop()
 
 	oneSecond := time.Second
 	KeepaliveInterval := time.Duration(0)
@@ -1300,11 +1300,11 @@ func TestConnectionStateFailedDeleteAllCandidates(t *testing.T) {
 
 // Assert that the ICE Agent can go directly from Connecting -> Failed on both sides
 func TestConnectionStateConnectingToFailed(t *testing.T) {
-	lim := test.TimeOut(time.Second * 5)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 5)
+	defer lim.Stop()
 
 	oneSecond := time.Second
 	KeepaliveInterval := time.Duration(0)
@@ -1361,11 +1361,11 @@ func TestConnectionStateConnectingToFailed(t *testing.T) {
 }
 
 func TestAgentRestart(t *testing.T) {
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
 
 	t.Run("Restart During Gather", func(t *testing.T) {
 		agent, err := NewAgent(&AgentConfig{})

--- a/agent_test.go
+++ b/agent_test.go
@@ -30,6 +30,9 @@ func (m *mockPacketConn) SetReadDeadline(t time.Time) error                   { 
 func (m *mockPacketConn) SetWriteDeadline(t time.Time) error                  { return nil }
 
 func TestPairSearch(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// Limit runtime in case of deadlocks
 	lim := test.TimeOut(time.Second * 10)
 	defer lim.Stop()
@@ -55,6 +58,9 @@ func TestPairSearch(t *testing.T) {
 }
 
 func TestPairPriority(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// avoid deadlocks?
 	defer test.TimeOut(1 * time.Second).Stop()
 
@@ -142,6 +148,9 @@ func TestPairPriority(t *testing.T) {
 }
 
 func TestOnSelectedCandidatePairChange(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// avoid deadlocks?
 	defer test.TimeOut(1 * time.Second).Stop()
 
@@ -220,7 +229,7 @@ func runAgentTest(t *testing.T, config *AgentConfig, task func(a *Agent)) {
 
 func TestHandlePeerReflexive(t *testing.T) {
 	// Limit runtime in case of deadlocks
-	lim := test.TimeOut(time.Second * 2)
+	lim := test.TimeOut(time.Second * 30)
 	defer lim.Stop()
 
 	report := test.CheckRoutines(t)

--- a/candidate.go
+++ b/candidate.go
@@ -37,6 +37,6 @@ type Candidate interface {
 
 	close() error
 	seen(outbound bool)
-	start(a *Agent, conn net.PacketConn)
+	start(a *Agent, conn net.PacketConn, initializedCh <-chan struct{})
 	writeTo(raw []byte, dst Candidate) (int, error)
 }

--- a/candidate_base.go
+++ b/candidate_base.go
@@ -72,19 +72,25 @@ func (c *candidateBase) RelatedAddress() *CandidateRelatedAddress {
 }
 
 // start runs the candidate using the provided connection
-func (c *candidateBase) start(a *Agent, conn net.PacketConn) {
+func (c *candidateBase) start(a *Agent, conn net.PacketConn, initializedCh <-chan struct{}) {
 	c.currAgent = a
 	c.conn = conn
 	c.closeCh = make(chan struct{})
 	c.closedCh = make(chan struct{})
 
-	go c.recvLoop()
+	go c.recvLoop(initializedCh)
 }
 
-func (c *candidateBase) recvLoop() {
+func (c *candidateBase) recvLoop(initializedCh <-chan struct{}) {
 	defer func() {
 		close(c.closedCh)
 	}()
+
+	select {
+	case <-initializedCh:
+	case <-c.closeCh:
+		return
+	}
 
 	log := c.agent().log
 	buffer := make([]byte, receiveMTU)

--- a/candidate_server_reflexive_test.go
+++ b/candidate_server_reflexive_test.go
@@ -14,12 +14,12 @@ import (
 )
 
 func TestServerReflexiveOnlyConnection(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// Limit runtime in case of deadlocks
 	lim := test.TimeOut(time.Second * 30)
 	defer lim.Stop()
-
-	report := test.CheckRoutines(t)
-	defer report()
 
 	serverPort := randomPort(t)
 	serverListener, err := net.ListenPacket("udp4", "127.0.0.1:"+strconv.Itoa(serverPort))

--- a/connectivity_vnet_test.go
+++ b/connectivity_vnet_test.go
@@ -430,11 +430,11 @@ func TestConnectivityVNet(t *testing.T) {
 
 // TestDisconnectedToConnected asserts that an agent can go to disconnected, and then return to connected successfully
 func TestDisconnectedToConnected(t *testing.T) {
-	lim := test.TimeOut(time.Second * 10)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 10)
+	defer lim.Stop()
 
 	loggerFactory := logging.NewDefaultLoggerFactory()
 
@@ -526,11 +526,11 @@ func TestDisconnectedToConnected(t *testing.T) {
 
 // Agent.Write should use the best valid pair if a selected pair is not yet available
 func TestWriteUseValidPair(t *testing.T) {
-	lim := test.TimeOut(time.Second * 10)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 10)
+	defer lim.Stop()
 
 	loggerFactory := logging.NewDefaultLoggerFactory()
 

--- a/gather_test.go
+++ b/gather_test.go
@@ -80,11 +80,11 @@ func TestListenUDP(t *testing.T) {
 
 // Assert that STUN gathering is done concurrently
 func TestSTUNConcurrency(t *testing.T) {
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
 
 	serverPort := randomPort(t)
 	serverListener, err := net.ListenPacket("udp4", "127.0.0.1:"+strconv.Itoa(serverPort))
@@ -139,11 +139,11 @@ func TestSTUNConcurrency(t *testing.T) {
 
 // Assert that TURN gathering is done concurrently
 func TestTURNConcurrency(t *testing.T) {
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
 
 	runTest := func(protocol ProtoType, scheme SchemeType, packetConn net.PacketConn, listener net.Listener, serverPort int) {
 		packetConnConfigs := []turn.PacketConnConfig{}

--- a/mdns_test.go
+++ b/mdns_test.go
@@ -12,12 +12,12 @@ import (
 )
 
 func TestMulticastDNSOnlyConnection(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// Limit runtime in case of deadlocks
 	lim := test.TimeOut(time.Second * 30)
 	defer lim.Stop()
-
-	report := test.CheckRoutines(t)
-	defer report()
 
 	cfg := &AgentConfig{
 		NetworkTypes:     []NetworkType{NetworkTypeUDP4},
@@ -54,12 +54,12 @@ func TestMulticastDNSOnlyConnection(t *testing.T) {
 }
 
 func TestMulticastDNSMixedConnection(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
 	// Limit runtime in case of deadlocks
 	lim := test.TimeOut(time.Second * 30)
 	defer lim.Stop()
-
-	report := test.CheckRoutines(t)
-	defer report()
 
 	aAgent, err := NewAgent(&AgentConfig{
 		NetworkTypes:     []NetworkType{NetworkTypeUDP4},
@@ -98,11 +98,11 @@ func TestMulticastDNSMixedConnection(t *testing.T) {
 }
 
 func TestMulticastDNSStaticHostName(t *testing.T) {
-	lim := test.TimeOut(time.Second * 30)
-	defer lim.Stop()
-
 	report := test.CheckRoutines(t)
 	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
 
 	_, err := NewAgent(&AgentConfig{
 		NetworkTypes:         []NetworkType{NetworkTypeUDP4},

--- a/transport.go
+++ b/transport.go
@@ -45,9 +45,6 @@ func (a *Agent) connect(ctx context.Context, isControlling bool, remoteUfrag, re
 	if err != nil {
 		return nil, err
 	}
-	if a.opened {
-		return nil, errors.New("a connection is already opened")
-	}
 	err = a.startConnectivityChecks(isControlling, remoteUfrag, remotePwd)
 	if err != nil {
 		return nil, err

--- a/transport_test.go
+++ b/transport_test.go
@@ -14,13 +14,13 @@ import (
 )
 
 func TestStressDuplex(t *testing.T) {
-	// Limit runtime in case of deadlocks
-	lim := test.TimeOut(time.Second * 20)
-	defer lim.Stop()
-
 	// Check for leaking routines
 	report := test.CheckRoutines(t)
 	defer report()
+
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 20)
+	defer lim.Stop()
 
 	// Run the test
 	stressDuplex(t)

--- a/transport_test.go
+++ b/transport_test.go
@@ -31,6 +31,13 @@ func testTimeout(t *testing.T, c *Conn, timeout time.Duration) {
 	const margin = 20 * time.Millisecond // allow 20msec error in time
 	statechan := make(chan ConnectionState, 1)
 	ticker := time.NewTicker(pollrate)
+	defer func() {
+		ticker.Stop()
+		err := c.Close()
+		if err != nil {
+			t.Error(err)
+		}
+	}()
 
 	startedAt := time.Now()
 
@@ -64,25 +71,37 @@ func TestTimeout(t *testing.T) {
 		t.Skip("skipping test in short mode.")
 	}
 
-	ca, cb := pipe(nil)
-	err := cb.Close()
+	// Check for leaking routines
+	report := test.CheckRoutines(t)
+	defer report()
 
-	if err != nil {
-		// we should never get here.
-		panic(err)
-	}
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 20)
+	defer lim.Stop()
 
-	testTimeout(t, ca, defaultDisconnectedTimeout)
+	t.Run("WithoutDisconnectTimeout", func(t *testing.T) {
+		ca, cb := pipe(nil)
+		err := cb.Close()
 
-	ca, cb = pipeWithTimeout(5*time.Second, 3*time.Second)
-	err = cb.Close()
+		if err != nil {
+			// we should never get here.
+			panic(err)
+		}
 
-	if err != nil {
-		// we should never get here.
-		panic(err)
-	}
+		testTimeout(t, ca, defaultDisconnectedTimeout)
+	})
 
-	testTimeout(t, ca, 5*time.Second)
+	t.Run("WithDisconnectTimeout", func(t *testing.T) {
+		ca, cb := pipeWithTimeout(5*time.Second, 3*time.Second)
+		err := cb.Close()
+
+		if err != nil {
+			// we should never get here.
+			panic(err)
+		}
+
+		testTimeout(t, ca, 5*time.Second)
+	})
 }
 
 func TestReadClosed(t *testing.T) {

--- a/transport_test.go
+++ b/transport_test.go
@@ -86,6 +86,14 @@ func TestTimeout(t *testing.T) {
 }
 
 func TestReadClosed(t *testing.T) {
+	// Check for leaking routines
+	report := test.CheckRoutines(t)
+	defer report()
+
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 20)
+	defer lim.Stop()
+
 	ca, cb := pipe(nil)
 
 	err := ca.Close()
@@ -351,6 +359,14 @@ func randomPort(t testing.TB) int {
 }
 
 func TestConnStats(t *testing.T) {
+	// Check for leaking routines
+	report := test.CheckRoutines(t)
+	defer report()
+
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 20)
+	defer lim.Stop()
+
 	ca, cb := pipe(nil)
 	if _, err := ca.Write(make([]byte, 10)); err != nil {
 		t.Fatal("unexpected error trying to write")

